### ISSUE TITLE
Freeze your script with Pyinstaller

### ISSUE THREAD

**Issue by masoudr** [ageitgey/face_recognition#357]
Hi,
You might want to freeze your script into a standalone executable to run on any computer without the need to install `python` or `face_recognition` and maybe you want to create a demo for your application and want to give it to someone else without giving your source code. Here is a simple tutorial to do that. I tested this method with `python3.6` on Windows 10 but I think you can get it working on Linux with similar method.
1. **Make sure** you have correctly installed both `face_recognition` and `dlib` correctly and you see no error when importing them into your script.
2. **Make sure** your script works fine and all it dependencies are right next to it and you can run it fine with `python yourscript.py`.
3. Install [Pyinstaller ](https://github.com/pyinstaller/pyinstaller)with `pip`:
`pip install pyinstaller`
4. Create a new directory and move your python script and all dependencies into it. I call it `myproject` and `myscript.py`
5. Copy `face_recognition_models` and `scipy-extra-dll` from your python installed directory to your project directory.
6. Create an empty file called `<yourscriptname>.spec` like `myscript.spec` next to your python script.
7. Use below Pyinstaller spec file sample and edit some parts according to your needs:
```
# -*- mode: python -*-

block_cipher = None

face_models = [
('.\\face_recognition_models\\models\\dlib_face_recognition_resnet_model_v1.dat', './face_recognition_models/models'),
('.\\face_recognition_models\\models\\mmod_human_face_detector.dat', './face_recognition_models/models'),
('.\\face_recognition_models\\models\\shape_predictor_5_face_landmarks.dat', './face_recognition_models/models'),
('.\\face_recognition_models\\models\\shape_predictor_68_face_landmarks.dat', './face_recognition_models/models'),
]

a = Analysis(['<your python script name.py>'],
             pathex=['<path to working directory>'],
             binaries=face_models,
             datas=[],
             hiddenimports=['scipy._lib.messagestream', 'scipy', 'scipy.signal', 'scipy.signal.bsplines', 'scipy.special', 'scipy.special._ufuncs_cxx',
                            'scipy.linalg.cython_blas',
                            'scipy.linalg.cython_lapack',
                            'scipy.integrate',
                            'scipy.integrate.quadrature',
                            'scipy.integrate.odepack',
                            'scipy.integrate._odepack',
                            'scipy.integrate.quadpack',
                            'scipy.integrate._quadpack',
                            'scipy.integrate._ode',
                            'scipy.integrate.vode',
                            'scipy.integrate._dop', 'scipy._lib', 'scipy._build_utils','scipy.__config__',
                            'scipy.integrate.lsoda', 'scipy.cluster', 'scipy.constants','scipy.fftpack','scipy.interpolate','scipy.io','scipy.linalg','scipy.misc','scipy.ndimage','scipy.odr','scipy.optimize','scipy.setup','scipy.sparse','scipy.spatial','scipy.special','scipy.stats','scipy.version'],

             hookspath=[],
             runtime_hooks=[],
             excludes=[],
             win_no_prefer_redirects=False,
             win_private_assemblies=False,
             cipher=block_cipher)

a.datas += Tree('./scipy-extra-dll', prefix=None)

pyz = PYZ(a.pure, a.zipped_data,
             cipher=block_cipher)
exe = EXE(pyz,
          a.scripts,
          a.binaries,
          a.zipfiles,
          a.datas,
          name='<your python script name>',
          debug=False,
          strip=False,
          upx=True,
          runtime_tmpdir=None,
          console=True )

```
8. Generate your executable with `python -m pyinstaller myscript.spec`
9. If you see no error then your executable can be found in `dist` directory.
10. Enjoy!

Thanks to @ageitgey and @davisking for their awesome work.

**Comment by ageitgey** [comment]
Cool, thanks for sharing!

**Comment by vgreis** [comment]
I found this worked very well when running on Windows desktop, but the application would crash on Windows Server 2012 - 2016 with an unhelpful, vague unable to execute DLL error. I found this issue was caused by using the latest version of dlib linked above (19.9.99 at this time). However, by using 'pip install dlib', which installs 19.9.0 (again, at this time). The frozen executable is running without errors on both desktop and server. I've also been testing freezing this in both 32-bit and 64-bit python and both are working. Thanks

**Comment by masoudr** [comment]
@vgreis I haven't tested Pyinstaller with the new version of `dlib`  but I think you can get it working by just add the missing DLL's to binaries it will then be extracted when running the app and again remember that it doesn't matter if you install `dlib` by source or `pip`, just make sure your installation is successful.

**Comment by mariocesarc** [comment]
Thanks! It works

**Comment by inewlife** [comment]
Incompatible library version: dlib.so requires version 51.0.0 or later, but libpng16.16.dylib provides version 38.0.0

**Comment by masoudr** [comment]
@inewlife Can you post the complete error log? I haven't tested this method with the new version of `dlib` but I think it is not related because what you need here is just adding `scipy` and `face_recognition_modules` manually not the `dlib` itself.

**Comment by fordffx** [comment]
I did as what the first floor told,but still got this:
Traceback (most recent call last):
  File "C:\PyInstaller-3.3.1\temp_classroom.py", line 16, in <module>
    import face_recognition
  File "<frozen importlib._bootstrap>", line 961, in _find_and_load
  File "<frozen importlib._bootstrap>", line 950, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 655, in _load_unlocked
  File "C:\PyInstaller-3.3.1\PyInstaller\loader\pyimod03_importers.py", line 631
, in exec_module
    exec(bytecode, module.__dict__)
  File "C:\ProgramData\Anaconda3\Lib\site-packages\face_recognition\__init__.py"
, line 7, in <module>
    from .api import load_image_file, face_locations, batch_face_locations, face
_landmarks, face_encodings, compare_faces, face_distance
  File "<frozen importlib._bootstrap>", line 961, in _find_and_load
  File "<frozen importlib._bootstrap>", line 950, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 655, in _load_unlocked
  File "C:\PyInstaller-3.3.1\PyInstaller\loader\pyimod03_importers.py", line 631
, in exec_module
    exec(bytecode, module.__dict__)
  File "C:\ProgramData\Anaconda3\Lib\site-packages\face_recognition\api.py", lin
e 21, in <module>
    pose_predictor_5_point = dlib.shape_predictor(predictor_5_point_model)
RuntimeError: Unable to open C:\Users\ADMINI~1\AppData\Local\Temp\_MEI71042\face
_recognition_models\models\shape_predictor_5_face_landmarks.dat
[5948] Failed to execute script temp_classroom

how did the  exe file come to this dir ? 
and _MEI71042 changes every time the exe file run

**Comment by masoudr** [comment]
@fordffx Hi,
As I said in the first post you must first install all dependencies for your python. About the `_MEI71042`, every time you run the executable `pyinstaller` tries to extract dependencies in the temp directory and when your program ends without exception it deletes that folder automatically,

**Comment by fordffx** [comment]
i replaced
datas=[]
with
datas=[('shape_predictor_68_face_landmarks.dat','./face_recognition_models/models'),('shape_predictor_5_face_landmarks.dat','./face_recognition_models/models'),('mmod_human_face_detector.dat','./face_recognition_models/models'),('dlib_face_recognition_resnet_model_v1.dat','./face_recognition_models/models')]
that works,thanks

**Comment by roiksail** [comment]
thanks for sharing your knowledge...i can't find scipy-extra-dll ...how i can to find them in python directory??i am in ubuntu 16.04,
thank you in advance

**Comment by masoudr** [comment]
@roiksail Hi, dll files are not made for Linux systems so you can't find any. I think you don't need to add `scipy` dependencies. Just build your binary file and try to run it. If It complains about some files just find that and add it manually like above method. That is it.

**Comment by roiksail** [comment]
thank you so much,i mmaked executable application in ubuntu but when i copied this file in ubuntu then it was simply file and it does not executed,how i fix this??

**Comment by masoudr** [comment]
@roiksail I think you need to find out what is Pyinstaller and how to use it in from [here](https://pythonhosted.org/PyInstaller/usage.html). In general, after you run the command you should find the executable with the exact name of the script. I think `Pyinstaller` will make the file as executable binary so you can just use `./yourscriptname` command and if it didn't work try to make it executable with `sudo chmod +x yourscriptname`. That is it.

**Comment by sincerefly** [comment]
@vgreis thanks, when I build use dlib from github source code, The version is 19.10.99, It can only run on x64 system, and when I use `pip install dlib`, the version is 19.10.0, It can run on x86 system

**Comment by sincerefly** [comment]
@masoudr

 I'm not find `scipy-extra-dll` in my site-packages, should I need install scipy? 

1, when I am not install scipy, I can run face_recognition, Is scipy necessary?
2, I install scipy, and copy scipy dir from site-packages to project dir, with `a.datas += Tree('./scipy-extra-dll', prefix=None)`, I can't find `scipy-extra-dll`, but I find `extra-dll` in scipy dir, can I write `scipy/extra-dll` instead it or not

thank you

**Comment by masoudr** [comment]
@sincerefly Hi, If I remember correctly there are some examples that need `scipy` so if you are using `scipy` with your project you need those files but if you don't,  just use `face_models` no need for `hidden imports` that is it. And remember you just need to first freeze your script normally and if you got any errors about missing packages, you need to feed them manually in your `spec` file.

**Comment by tinshade** [comment]
@masoudr Can you please explain to me how can I get the face_recognition_models/**/*.dat resources added from the .spec file? It does not seem to work for me. It says unable to open shape_predictor_68_face_landmarks.dat for some reason.

So far, I have tried:

-  Creating with --onedir

-  Adding the files to the directory and then running pyinstaller with --onedir.

**Comment by masoudr** [comment]
@tinshade I've already added all 4 model files in the example above, you just need to use the above sample spec file and fill parameters and this should solve your problem. But if you still need to add other models you can just add this to your spec file.

```
a = Analysis(...
     datas= [ ('face_recognition_models/**/*.dat', './face_recognition_models/models' ) ],
     ...
     )
```

**Comment by alexpardede** [comment]
i already install Pyinstaller, and while i check the directory its true no module name Pyinstaller. when i check in /home/attendance/anaconda3/envs/attendance_env/bin/python3.6/site-packages i find the module.

```
(attendance_env) attendance@attendance-desktop:~/Documents/face_attendance$ python -m Pyinstaller attendance.spec
/home/attendance/anaconda3/envs/attendance_env/bin/python: No module named Pyinstaller

```

can u help me

**Comment by masoudr** [comment]
@alexpardede I'm not familiar with `Conda` but I would suggest you first check your naming, it is case sensitive (`python -m PyInstaller`). Second, you can check your installed modules with `pip freeze`. Also if you have any problem with installing `PyInstaller` you should ask in its [repo](https://github.com/pyinstaller/pyinstaller).

**Comment by alexpardede** [comment]
thansk for your help. python -m PyInstaller work for me. but i didn't
solve the step 5 cause i don't know where i can find the module
[image: image.png]
[image: image.png]
and in step 5. how to copy face_recognition models to project directory. i
didn't find the module. can u help me to solve this step ?
[image: image.png]

thank you for your help.

Pada tanggal Jum, 29 Mar 2019 pukul 00.41 Masoud R. <
notifications@github.com> menulis:

**Comment by alexpardede** [comment]
![image](https://user-images.githubusercontent.com/43948231/55208909-a86b3d00-5212-11e9-98ad-a0cbe52b89c1.png)

can you help me for this error ? i already have face-recognition-models while i check with pip freeze. and i already create and execute the attendance.spec. please help me. Thank You

**Comment by zkywalker** [comment]
> ![image](https://user-images.githubusercontent.com/43948231/55208909-a86b3d00-5212-11e9-98ad-a0cbe52b89c1.png)
> 
> can you help me for this error ? i already have face-recognition-models while i check with pip freeze. and i already create and execute the attendance.spec. please help me. Thank You

try to copy the dependencies that open fail to your output dependencies dir, its helpful for me

**Comment by alexpardede** [comment]
thanks. i already done with this problem.
thanks for your attention

Pada tanggal Jum, 29 Mar 2019 pukul 13.40 zkywalker <
notifications@github.com> menulis:

**Comment by alexpardede** [comment]
![image](https://user-images.githubusercontent.com/43948231/55227664-aae98900-524a-11e9-83e9-ecf4a1f0f4f7.png)
i have new issue.  my main script refere to a other script. and the script call library "frozen_inference_graph_face.pb" from another folder. can you help me to solve the problem  ?
thank for your attention.

**Comment by zkywalker** [comment]
> ![image](https://user-images.githubusercontent.com/43948231/55227664-aae98900-524a-11e9-83e9-ecf4a1f0f4f7.png)
> i have new issue. my main script refere to a other script. and the script call library "frozen_inference_graph_face.pb" from another folder. can you help me to solve the problem ?
> thank for your attention.

lol, the solution is same to dependencies no find, just copy them to the output dependencies dir. And its helpful when you do not use the -f arg, which will generate a file with all dependencies in it.

**Comment by alexpardede** [comment]
sorry brother. but i already copy to the output dependencies dir. and i'm
not using -f  arg. if the file in the different dependencies i don't show
the issue.

Pada tanggal Jum, 29 Mar 2019 pukul 18.37 zkywalker <
notifications@github.com> menulis:

**Comment by alexpardede** [comment]
[image: image.png]
can you help me with this problem ?

[image: image.png]
i already insert the library. please help me
thank you

Pada tanggal Jum, 29 Mar 2019 pukul 00.41 Masoud R. <
notifications@github.com> menulis:

**Comment by sandipJava** [comment]
Hello I am facing following issue
![error1](https://user-images.githubusercontent.com/32532237/65404019-63584a00-ddf4-11e9-9cbd-bd9eeb3a76e3.jpg)

**Comment by libohao666** [comment]
> I found this worked very well when running on Windows desktop, but the application would crash on Windows Server 2012 - 2016 with an unhelpful, vague unable to execute DLL error. I found this issue was caused by using the latest version of dlib linked in the project install guide under issue 175 (currently 19.9.99). However, by using 'pip install dlib', which installs 19.9.0 (again, currently as of writing this). The frozen executable is running without errors on both desktop and server. I've also been testing freezing this in both 32-bit and 64-bit python and both are working. Thanks

sorry to bother you.I don't understand.is 19.9.0dlib installed in the computer which pyinstaller the file or in winserver2012?My my application can run in my windows computer,but it can't work on my windows server2012.And my dlib package can't be uninstalled completely now....

**Comment by fordffx** [comment]
> > I found this worked very well when running on Windows desktop, but the application would crash on Windows Server 2012 - 2016 with an unhelpful, vague unable to execute DLL error. I found this issue was caused by using the latest version of dlib linked in the project install guide under issue 175 (currently 19.9.99). However, by using 'pip install dlib', which installs 19.9.0 (again, currently as of writing this). The frozen executable is running without errors on both desktop and server. I've also been testing freezing this in both 32-bit and 64-bit python and both are working. Thanks
> 
> sorry to bother you.I don't understand.is 19.9.0dlib installed in the computer which pyinstaller the file or in winserver2012?My my application can run in my windows computer,but it can't work on my windows server2012.And my dlib package can't be uninstalled completely now....

maybe you could try building the latest version of dlib on the target platform

**Comment by Ragesh77** [comment]
> 31 INFO: PyInstaller: 3.5
> 32 INFO: Python: 3.6.9
> 32 INFO: Platform: Linux-5.0.0-37-generic-x86_64-with-Zorin-15-bionic
> 33 INFO: UPX is not available.
> 35 INFO: Extending PYTHONPATH with paths
> ['/home/ragesh/Desktop/myproject', '/home/ragesh/Desktop/myproject']
> 35 INFO: checking Analysis
> 35 INFO: Building Analysis because Analysis-00.toc is non existent
> 35 INFO: Initializing module dependency graph...
> 36 INFO: Initializing module graph hooks...
> 37 INFO: Analyzing base_library.zip ...
> 2303 INFO: Analyzing hidden import 'scipy._lib.messagestream'
> 2704 INFO: Processing pre-find module path hook   distutils
> 3588 INFO: Processing pre-safe import module hook   setuptools.extern.six.moves
> 3886 INFO: Processing pre-find module path hook   site
> 3886 INFO: site: retargeting to fake-dir '/home/ragesh/.local/lib/python3.6/site-packages/PyInstaller/fake-modules'
> 6109 INFO: Analyzing hidden import 'scipy.signal'
> 7681 INFO: Processing pre-safe import module hook   six.moves
> 11696 INFO: Processing pre-safe import module hook   gi.repository.GObject
> 13029 INFO: Analyzing hidden import 'scipy.special._ufuncs_cxx'
> 13029 INFO: Analyzing hidden import 'scipy.linalg.cython_blas'
> 13029 INFO: Analyzing hidden import 'scipy.linalg.cython_lapack'
> 13029 INFO: Analyzing hidden import 'scipy._build_utils'
> 13050 INFO: Analyzing hidden import 'scipy.cluster'
> 13126 INFO: Analyzing hidden import 'scipy.fftpack'
> 13157 INFO: Analyzing hidden import 'scipy.io'
> 13332 INFO: Analyzing hidden import 'scipy.odr'
> 13370 INFO: Analyzing hidden import 'scipy.setup'
> 13374 INFO: running Analysis Analysis-00.toc
> 13399 INFO: Caching module hooks...
> 13402 INFO: Analyzing script.py
> 13455 INFO: Loading module hooks...
> 13455 INFO: Loading module hook "hook-scipy.py"...
> 13456 INFO: Loading module hook "hook-PyQt5.QtGui.py"...
> 13537 INFO: Loading module hook "hook-PyQt5.QtCore.py"...
> 13556 INFO: Loading module hook "hook-pytz.py"...
> 13557 INFO: Loading module hook "hook-scipy.io.matlab.py"...
> 13557 INFO: Loading module hook "hook-PyQt5.py"...
> 15933 WARNING: Hidden import "PyQt5.sip" not found!
> 15933 INFO: Loading module hook "hook-scipy.sparse.csgraph.py"...
> 15935 INFO: Loading module hook "hook-PIL.Image.py"...
> 16323 INFO: Loading module hook "hook-numpy.core.py"...
> 16450 INFO: Loading module hook "hook-matplotlib.backends.py"...
> 16934 INFO:   Matplotlib backend "GTK": ignored
>     Gtk* backend requires pygtk to be installed.
> 17828 INFO:   Matplotlib backend "GTKAgg": ignored
>     Gtk* backend requires pygtk to be installed.
> 18071 INFO:   Matplotlib backend "GTKCairo": ignored
>     No module named 'gtk'
> 18409 INFO:   Matplotlib backend "MacOSX": ignored
>     cannot import name '_macosx'
> 18768 INFO:   Matplotlib backend "Qt4Agg": added
> 19143 INFO:   Matplotlib backend "Qt5Agg": added
> 19489 INFO:   Matplotlib backend "TkAgg": added
> 19828 INFO:   Matplotlib backend "WX": ignored
>     Matplotlib backend_wx and backend_wxagg require wxPython >=2.8.12
> 20167 INFO:   Matplotlib backend "WXAgg": ignored
>     Matplotlib backend_wx and backend_wxagg require wxPython >=2.8.12
> 20577 INFO:   Matplotlib backend "GTK3Cairo": added
> /usr/lib/python3/dist-packages/matplotlib/backends/backend_gtk3agg.py:17: UserWarning: The Gtk3Agg backend is known to not work on Python 3.x with pycairo. Try installing cairocffi.
>   "The Gtk3Agg backend is known to not work on Python 3.x with pycairo. "
> 20998 INFO:   Matplotlib backend "GTK3Agg": added
> 21604 INFO:   Matplotlib backend "WebAgg": added
> 21889 INFO:   Matplotlib backend "nbAgg": ignored
>     No module named 'IPython'
> 22229 INFO:   Matplotlib backend "agg": added
> 22570 INFO:   Matplotlib backend "cairo": added
> 22813 INFO:   Matplotlib backend "gdk": ignored
>     No module named 'gobject'
> 23129 INFO:   Matplotlib backend "pdf": added
> 23491 INFO:   Matplotlib backend "pgf": added
> 23784 INFO:   Matplotlib backend "ps": added
> 24135 INFO:   Matplotlib backend "svg": added
> 24474 INFO:   Matplotlib backend "template": added
> 24651 INFO: Processing pre-safe import module hook   gi.repository.Gtk
> 24651 INFO: Processing pre-safe import module hook   gi.repository.Gdk
> 24651 INFO: Processing pre-safe import module hook   gi.repository.GLib
> 24749 INFO: Loading module hook "hook-pkg_resources.py"...
> 25087 INFO: Processing pre-safe import module hook   win32com
> 25093 INFO: Loading module hook "hook-distutils.py"...
> 25093 INFO: Loading module hook "hook-matplotlib.py"...
> 25337 INFO: Loading module hook "hook-pydoc.py"...
> 25338 INFO: Loading module hook "hook-setuptools.py"...
> 25761 INFO: Loading module hook "hook-gi.repository.GObject.py"...
> 25898 WARNING: Hidden import "gi._gobject" not found!
> 25898 INFO: Loading module hook "hook-lib2to3.py"...
> 25899 INFO: Loading module hook "hook-_tkinter.py"...
> 25974 INFO: checking Tree
> 25974 INFO: Building Tree because Tree-00.toc is non existent
> 25974 INFO: Building Tree Tree-00.toc
> 25979 INFO: checking Tree
> 25979 INFO: Building Tree because Tree-01.toc is non existent
> 25979 INFO: Building Tree Tree-01.toc
> 25981 INFO: Loading module hook "hook-gi.py"...
> 25981 INFO: Loading module hook "hook-scipy.linalg.py"...
> 25982 INFO: Loading module hook "hook-encodings.py"...
> 26029 INFO: Loading module hook "hook-gi.repository.Gtk.py"...
> 28333 INFO: Processing pre-safe import module hook   gi.repository.xlib
> 28334 INFO: Processing pre-safe import module hook   gi.repository.Atk
> 28334 INFO: Loading module hook "hook-scipy.special._ufuncs.py"...
> 28334 INFO: Loading module hook "hook-PIL.py"...
> 28337 INFO: Excluding import 'tkinter'
> 28339 INFO:   Removing import of tkinter from module PIL.ImageTk
> 28339 INFO: Import to be excluded not found: 'FixTk'
> 28340 INFO: Excluding import 'PyQt5'
> 28341 INFO:   Removing import of PyQt5.QtGui from module PIL.ImageQt
> 28341 INFO:   Removing import of PyQt5.QtCore from module PIL.ImageQt
> 28342 INFO: Excluding import 'PyQt4'
> 28344 INFO: Excluding import 'PySide'
> 28347 INFO: Loading module hook "hook-PyQt5.QtWidgets.py"...
> 28411 INFO: Loading module hook "hook-scipy.special._ellip_harm_2.py"...
> 28411 INFO: Loading module hook "hook-xml.etree.cElementTree.py"...
> 28412 INFO: Loading module hook "hook-cv2.py"...
> 28412 INFO: Loading module hook "hook-numpy.py"...
> 28412 INFO: Loading module hook "hook-sysconfig.py"...
> 28418 INFO: Loading module hook "hook-gi.repository.GLib.py"...
> 28610 INFO: Loading module hook "hook-xml.py"...
> 28623 INFO: Loading module hook "hook-PIL.SpiderImagePlugin.py"...
> 28627 INFO: Excluding import 'tkinter'
> 28628 INFO: Import to be excluded not found: 'FixTk'
> 28628 INFO: Loading module hook "hook-gi.repository.Gdk.py"...
> 28741 INFO: Processing pre-safe import module hook   gi.repository.cairo
> 28742 INFO: Processing pre-safe import module hook   gi.repository.Pango
> 28743 INFO: Processing pre-safe import module hook   gi.repository.Gio
> 28743 INFO: Processing pre-safe import module hook   gi.repository.GdkPixbuf
> 28743 INFO: Loading module hook "hook-gi.repository.GdkPixbuf.py"...
> 28744 WARNING: "hook-gi.repository.GdkPixbuf" ignored, since GDK not found (i.e., "gdk-pixbuf-query-loaders" not in $PATH).
> 28744 INFO: Loading module hook "hook-gi.repository.xlib.py"...
> 28840 INFO: Loading module hook "hook-gi.repository.cairo.py"...
> 28944 INFO: Loading module hook "hook-gi.repository.Pango.py"...
> 29049 INFO: Loading module hook "hook-gi.repository.Atk.py"...
> 29155 INFO: Loading module hook "hook-gi.repository.Gio.py"...
> 29617 INFO: Looking for ctypes DLLs
> 29718 WARNING: library user32 required via ctypes not found
> 29754 INFO: Analyzing run-time hooks ...
> 29766 INFO: Including run-time hook 'pyi_rth_mplconfig.py'
> 29768 INFO: Including run-time hook 'pyi_rth_mpldata.py'
> 29768 INFO: Including run-time hook 'pyi_rth_glib.py'
> 29769 INFO: Including run-time hook 'pyi_rth_gdkpixbuf.py'
> 29771 INFO: Including run-time hook 'pyi_rth_gio.py'
> 29771 INFO: Including run-time hook 'pyi_rth_gtk.py'
> 29772 INFO: Including run-time hook 'pyi_rth__tkinter.py'
> 29773 INFO: Including run-time hook 'pyi_rth_pyqt5.py'
> 29774 INFO: Including run-time hook 'pyi_rth_gi.py'
> 29776 INFO: Including run-time hook 'pyi_rth_multiprocessing.py'
> 29779 INFO: Including run-time hook 'pyi_rth_pkgres.py'
> 29797 INFO: Looking for dynamic libraries
> 34474 INFO: Looking for eggs
> 34474 INFO: Python library not in binary dependencies. Doing additional searching...
> 34499 INFO: Using Python library /usr/lib/x86_64-linux-gnu/libpython3.6m.so.1.0
> 34538 INFO: Warnings written to /home/ragesh/Desktop/myproject/build/script/warn-script.txt
> 34637 INFO: Graph cross-reference written to /home/ragesh/Desktop/myproject/build/script/xref-script.html
> 36326 INFO: Appending 'binaries' from .spec
> Unable to find ".\face_recognition_models\models\dlib_face_recognition_resnet_model_v1.dat" when adding binary and data files.
> 

Help me with this.

**Comment by Ragesh77** [comment]
I can't find the **scipy-extra-dll**  in windows although I installed scipy. Help me.

**Comment by masoudr** [comment]
@Ragesh77 I haven't tested the new version. But check if you can import `scipy` without any problem and then double-check the file existence. If you still can't find it skip it and check if the executable works or not.

**Comment by Ragesh77** [comment]
@masoudr Hey Than

> @Ragesh77 I haven't tested the new version. But check if you can import `scipy` without any problem and then double-check the file existence. If you still can't find it skip it and check if the executable works or not.
@masoudr  Thanks for answering. I downgraded scipy to 1.3.0 and could locate all the files.

**Comment by HAKANMAZI** [comment]
> I can't find the **scipy-extra-dll** in windows although I installed scipy. Help me.

You can search extra-dll using below code on cmd
dir extra-dll /s /p
I hope you will find it in scipy directory. After create scipy-extra-dll directory. 
Then all dll files transfer to scipy-extra-dll directory.

**Comment by HAKANMAZI** [comment]
> Hi,
> You might want to freeze your script into a standalone executable to run on any system without the need of installing `python` or `face_recognition` and maybe you want to create a demo for your application and want to give it to someone else without giving your source code. Here is a simple tutorial to do that. I tested this method with `python3.6` on Windows 10 but I think you can get it working on Linux with similar method.
> 
> 1. **Make sure** you have correctly installed both `face_recognition` and `dlib` correctly and you see no error when importing them into your script.
> 2. **Make sure** your script works fine and all it dependencies are right next to it and you can run it fine with `python yourscript.py`.
> 3. Install [Pyinstaller ](https://github.com/pyinstaller/pyinstaller)with `pip`:
>    `pip install pyinstaller`
> 4. Create a new directory and move your python script and all dependencies into it. I call it `myproject` and `myscript.py`
> 5. Copy `face_recognition_models` and `scipy-extra-dll` from your python installed directory to your project directory.
> 6. Create an empty file called `<yourscriptname>.spec` like `myscript.spec` next to your python script.
> 7. Use below Pyinstaller spec file sample and edit some parts according to your needs: (I mark it with `<>` tag)
> 
> ```
> # -*- mode: python -*-
> 
> block_cipher = None
> 
> face_models = [
> ('.\\face_recognition_models\\models\\dlib_face_recognition_resnet_model_v1.dat', './face_recognition_models/models'),
> ('.\\face_recognition_models\\models\\mmod_human_face_detector.dat', './face_recognition_models/models'),
> ('.\\face_recognition_models\\models\\shape_predictor_5_face_landmarks.dat', './face_recognition_models/models'),
> ('.\\face_recognition_models\\models\\shape_predictor_68_face_landmarks.dat', './face_recognition_models/models'),
> ]
> 
> a = Analysis(['<your python script name.py>'],
>              pathex=['<path to working directory>'],
>              binaries=face_models,
>              datas=[],
>              hiddenimports=['scipy._lib.messagestream', 'scipy', 'scipy.signal', 'scipy.signal.bsplines', 'scipy.special', 'scipy.special._ufuncs_cxx',
>                             'scipy.linalg.cython_blas',
>                             'scipy.linalg.cython_lapack',
>                             'scipy.integrate',
>                             'scipy.integrate.quadrature',
>                             'scipy.integrate.odepack',
>                             'scipy.integrate._odepack',
>                             'scipy.integrate.quadpack',
>                             'scipy.integrate._quadpack',
>                             'scipy.integrate._ode',
>                             'scipy.integrate.vode',
>                             'scipy.integrate._dop', 'scipy._lib', 'scipy._build_utils','scipy.__config__',
>                             'scipy.integrate.lsoda', 'scipy.cluster', 'scipy.constants','scipy.fftpack','scipy.interpolate','scipy.io','scipy.linalg','scipy.misc','scipy.ndimage','scipy.odr','scipy.optimize','scipy.setup','scipy.sparse','scipy.spatial','scipy.special','scipy.stats','scipy.version'],
> 
>              hookspath=[],
>              runtime_hooks=[],
>              excludes=[],
>              win_no_prefer_redirects=False,
>              win_private_assemblies=False,
>              cipher=block_cipher)
> 
> a.datas += Tree('./scipy-extra-dll', prefix=None)
> 
> pyz = PYZ(a.pure, a.zipped_data,
>              cipher=block_cipher)
> exe = EXE(pyz,
>           a.scripts,
>           a.binaries,
>           a.zipfiles,
>           a.datas,
>           name='<your python script name>',
>           debug=False,
>           strip=False,
>           upx=True,
>           runtime_tmpdir=None,
>           console=True )
> ```
> 
> 1. Generate your executable with `python -m pyinstaller myscript.spec`
> 2. If you see no error then your executable can be found in `dist` directory.
> 3. Enjoy!
> 
> Thanks to @ageitgey and @davisking for their awesome work.

Thanks for everything all guys. I did what you said and I created an exe file successfully. But when I click exe file it's working only once then suddenly closes. Before close it says something but I don't read because it is so fastly close. How can I fix this problem

**Comment by masoudr** [comment]
@HAKANMAZI Try to run executable with CMD and you should get the error in the console. Then put the error.

**Comment by Ragesh77** [comment]
Can Someone help me in deploying this app in docker.I searched google nothing is clear.Could someone guide me?

**Comment by masoudr** [comment]
@Ragesh77  It is not an app. But if you trying to deploy your app with docker you can follow [this](https://github.com/ageitgey/face_recognition#deployment).

**Comment by HAKANMAZI** [comment]
> @HAKANMAZI Try to run executable with CMD and you should get the error in the console. Then put the error.

Thanks @masoudr ;  output error below


T

> raceback (most recent call last):
>   File "C:\Users\HH\Desktop\python\exe\test.py", line 94, in <module>
>     openCamera()
>   File "C:\Users\HH\Desktop\python\exe\test.py", line 66, in openCamera
>     predictions = predict(facePath, model_path=model_path)
>   File "C:\Users\HH\Desktop\python\exe\test.py", line 30, in predict
>     knn_clf = pickle.load(f)
> ModuleNotFoundError: No module named 'sklearn'
> [14372] Failed to execute script test
> [ WARN:0] global C:\projects\opencv-python\opencv\modules\videoio\src\cap_msmf.cpp (674) SourceReaderCB::~SourceReaderCB terminating async callback

**Comment by masoudr** [comment]
@HAKANMAZI Your error is not related to this tutorial (because you are using `sklearn` which is another module), but you can use [this](https://stackoverflow.com/questions/57108026/pyinstaller-modulenotfounderror-no-module-named-sklearn-utils-cython-blas/57108887#57108887).

**Comment by HAKANMAZI** [comment]
> @HAKANMAZI Your error is not related to this tutorial (because you are using `sklearn` which is another module), but you can use [this](https://stackoverflow.com/questions/57108026/pyinstaller-modulenotfounderror-no-module-named-sklearn-utils-cython-blas/57108887#57108887).

But I don't use sklearn module.
Here is my ScreenShot_FacePredict.py;

> import face_recognition
> import pickle
> import cv2
> import os
> import pyttsx3
> from PIL import ImageGrab,Image
> 
> 
> 
> model_path=r"C:\Users\HH\Desktop\ekranFace\yabanci_knn_unlu_970.clf"
> 
> def konus():
>     engine = pyttsx3.init()
>     engine.say(" Bulundu ")
>     engine.runAndWait()
> 
> def predict(x_img_path, knn_clf=None, model_path=None, distance_threshold=0.47):
>     if knn_clf is None:
>         with open(model_path, 'rb') as f:
>             knn_clf = pickle.load(f)
>             
>     x_img = face_recognition.load_image_file(x_img_path)
>     x_face_locations = face_recognition.face_locations(x_img,number_of_times_to_upsample=1)
>     
>     if len(x_face_locations) == 0:
>         return []
>         
>     faces_encodings = face_recognition.face_encodings(x_img, known_face_locations=x_face_locations,num_jitters=2)
>     
>     closest_distances = knn_clf.kneighbors(faces_encodings, n_neighbors=1)
>     #print(closest_distances)
>     are_matches = [closest_distances[0][i][0] <= distance_threshold for i in range(len(x_face_locations))]
>     
>     return [(pred, loc) if rec else ("bilinmiyor", loc) for pred, loc, rec in zip(knn_clf.predict(faces_encodings), x_face_locations, are_matches)]
>  
> 
> def ekranGrountusuAl():
>     resim = ImageGrab.grab()                  #Get ScreenShot
>     resim.save("ekran.png",'png')                 #ScreenShot name and save
>     test_dir = os.getcwd()                    #Get py path 
>     
>     name2="name2"
>     counter =0
>     while True:
>         ekran = face_recognition.load_image_file(test_dir+"\\"+"ekran.png") 
>         face_locations = face_recognition.face_locations(ekran)
>         
>         if len(face_locations)>=1:
>             for face_location in face_locations:
>                 top, right, bottom, left = face_location
>                 face = ekran[top:bottom, left:right]
>                 
>                 #Save Face
>                 facePath = test_dir+"\\"+"face.png"
>                 cv2.imwrite(facePath, face)
>                 
>                 #Show Face
>                 size = 512, 512
>                 image = Image.open(facePath)
>                 image.thumbnail(size)
>                 image.show()
>                 
>                 
>                 predictions = predict(facePath, model_path=model_path)
>                 for name, (top, right, bottom, left) in predictions:
>                     print("- Yuz {}".format(name))
>                     if name == name2:
>                         counter = 1
>                     
>                     if name != "bilinmiyor":
>                         name2 = name
>                 
>                 
>             if counter != 0:
>                 print(f"{name2}  bulundu")
>                 konus()
>                 break
>         ekranGrountusuAl()
> 
>         
> if __name__ == "__main__":
>     ekranGrountusuAl()
      
    

Here is my test.spec file
    
> # -*- mode: python -*-
> 
> block_cipher = None
> 
> face_models = [
> ('.\\face_recognition_models\\models\\dlib_face_recognition_resnet_model_v1.dat', './face_recognition_models/models'),
> ('.\\face_recognition_models\\models\\mmod_human_face_detector.dat', './face_recognition_models/models'),
> ('.\\face_recognition_models\\models\\shape_predictor_5_face_landmarks.dat', './face_recognition_models/models'),
> ('.\\face_recognition_models\\models\\shape_predictor_68_face_landmarks.dat', './face_recognition_models/models'),
> ]
> 
> a = Analysis(['ScreenShot_FacePredict.py'],
>              pathex=[r'C:\Users\HH\Desktop\ekranFace'],
>              binaries=face_models,
>              datas=[],
>              hiddenimports=['scipy._lib.messagestream',
> 							'scipy', 'scipy.signal', 'scipy.signal.bsplines', 'scipy.special', 'scipy.special._ufuncs_cxx',
>                             'scipy.linalg.cython_blas',
>                             'scipy.linalg.cython_lapack',
>                             'scipy.integrate',
>                             'scipy.integrate.quadrature',
>                             'scipy.integrate.odepack',
>                             'scipy.integrate._odepack',
>                             'scipy.integrate.quadpack',
>                             'scipy.integrate._quadpack',
>                             'scipy.integrate._ode',
>                             'scipy.integrate.vode',
>                             'scipy.integrate._dop', 'scipy._lib', 'scipy._build_utils','scipy.__config__',
>                             'scipy.integrate.lsoda', 'scipy.cluster', 'scipy.constants','scipy.fftpack','scipy.interpolate','scipy.io','scipy.linalg','scipy.misc','scipy.ndimage','scipy.odr','scipy.optimize','scipy.setup','scipy.sparse','scipy.spatial','scipy.special','scipy.stats','scipy.version'],
> 
>              hookspath=[],
>              runtime_hooks=[],
>              excludes=[],
>              win_no_prefer_redirects=False,
>              win_private_assemblies=False,
>              cipher=block_cipher)
> 
> a.datas += Tree('./scipy-extra-dll', prefix=None)
> 
> pyz = PYZ(a.pure, a.zipped_data,
>              cipher=block_cipher)
> exe = EXE(pyz,
>           a.scripts,
>           a.binaries,
>           a.zipfiles,
>           a.datas,
>           name='test',
>           debug=False,
>           strip=False,
>           upx=True,
>           runtime_tmpdir=None,
>           console=True )
> 

After all doing necessary things I created successfully exe file. But it gives me this error.

> Traceback (most recent call last):
>   File "C:\Users\HH\Desktop\ekranFace\ScreenShot_FacePredict.py", line 82, in <module>
>     ekranGrountusuAl()
>   File "C:\Users\HH\Desktop\ekranFace\ScreenShot_FacePredict.py", line 64, in ekranGrountusuAl
>     predictions = predict(facePath, model_path=model_path)
>   File "C:\Users\HH\Desktop\ekranFace\ScreenShot_FacePredict.py", line 20, in predict
>     knn_clf = pickle.load(f)
> ModuleNotFoundError: No module named 'sklearn'
> [8864] Failed to execute script ScreenShot_FacePredict

I don't use sklearn as you see on upside. So How can ı solve this promblem?

**Comment by HAKANMAZI** [comment]
> > @HAKANMAZI Try to run executable with CMD and you should get the error in the console. Then put the error.
> 
> Thanks @masoudr ; output error below, error says me No module named 'sklearn'. Yes true there is no sklearn because I don't use it. Even if when I run my .py file there is no error. But when I change from .py to .exe file it saying No module named 'sklearn'. Why ?
> 
> > raceback (most recent call last):
> > File "C:\Users\HH\Desktop\python\exe\test.py", line 94, in 
> > openCamera()
> > File "C:\Users\HH\Desktop\python\exe\test.py", line 66, in openCamera
> > predictions = predict(facePath, model_path=model_path)
> > File "C:\Users\HH\Desktop\python\exe\test.py", line 30, in predict
> > knn_clf = pickle.load(f)
> > ModuleNotFoundError: No module named 'sklearn'
> > [14372] Failed to execute script test
> > [ WARN:0] global C:\projects\opencv-python\opencv\modules\videoio\src\cap_msmf.cpp (674) SourceReaderCB::~SourceReaderCB terminating async callback

Finally I solved my problem. 
As I said I don't use sklearn on my code but background sklearn was using. Then I add in hiddenimports=['sklearn','sklearn.neighbors._typedefs','sklearn.utils._cython_blas'] in my spec file. 
and it is okey, it solved :)

**Comment by daynial132** [comment]
hello I am new to this freezer part,

I have made a separate folder as mentioned and move my python script and models in it. 
also i copied the freezer script(.spec) file. finally i got the build giving the error "ImportError: No module named 'face_recognition'"
Please help me with it I am using anaconda, and not using scipy. 
I have attached the freezer script also

[freeze_app .txt](https://github.com/ageitgey/face_recognition/files/4282895/freeze_app.txt)

**Comment by masoudr** [comment]
@daynial132 First make sure that your script is working fine in your environment also you must be sure that you run PyInstaller from your environment (with using `python -m PyInstaller`). I think you are running the PyInstaller from your main Python installation which doesn't have face_recognition.

**Comment by daynial132** [comment]
@masoudr I have placed my python script,3 pics and the freezer file (.spec) and the face_recognition_models in the folder only. do I need any thing else?

I have check my python script to run on my anaconda shell, it is running fine that's mean dlib and face_recognition lib is installed properly. Am i right or missing some thing?

Also I have used anaconda shell to run (python -m PyInstaller) command. 

Thank you for your reply

**Comment by masoudr** [comment]
@daynial132 I haven't tested this with Conda yet, but I don't think it would be a problem. Your error means that PyInstaller can't find the `face_recognition`. Try to run `pip freeze` inside your console and make sure that face_recognition exists in the output. If it is there but you still get that error try to add `face_recognition` into the `hidden_imports` as I did for the scipy modules.

**Comment by mend4x** [comment]
I am having this output after freezing my code with this `.spec` file below. When I run my app on console on Linux I get this:
 
```
Traceback (most recent call last):
  File "main.py", line 6, in <module>
  File "<frozen importlib._bootstrap>", line 971, in _find_and_load
  File "<frozen importlib._bootstrap>", line 955, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 665, in _load_unlocked
  File "/home/neo/Desktop/faseez/dev/refasee/venv/lib/python3.6/site-packages/PyInstaller/loader/pyimod03_importers.py", line 623, in exec_module
    exec(bytecode, module.__dict__)
  File "face_recognition/__init__.py", line 7, in <module>
  File "<frozen importlib._bootstrap>", line 971, in _find_and_load
  File "<frozen importlib._bootstrap>", line 955, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 665, in _load_unlocked
  File "/home/neo/Desktop/faseez/dev/refasee/venv/lib/python3.6/site-packages/PyInstaller/loader/pyimod03_importers.py", line 623, in exec_module
    exec(bytecode, module.__dict__)
  File "face_recognition/api.py", line 20, in <module>
RuntimeError: Unable to open /tmp/_MEIHAOamS/face_recognition_models/models/shape_predictor_68_face_landmarks.dat
[11437] Failed to execute script main
```


[Fascee.spec.txt](https://github.com/ageitgey/face_recognition/files/4382857/Fascee.spec.txt)

**Comment by masoudr** [comment]
@mend4x Make sure that you have the `shape_predictor_68_face_landmarks.dat` when building the project. Also if the error came double check that the `shape_predictor_68_face_landmarks.dat` exists in the given path (e.g `/tmp/_MEIHAOamS/face_recognition_models/models`).

**Comment by mend4x** [comment]
@masoudr I actually do have this file in script directory. The given path is `tmp` so each time it runs it's new directory

**Comment by masoudr** [comment]
@mend4x Double check that the files in `models` directory and also when you run the script check the `tmp` directory and verify the file and its size.

**Comment by tingfengyinyue** [comment]
@masoudr   我的问题是，打包的时候没有问题，打包后，点击exe 只有小黑窗口，程序没办法运行。
界面没办法显示。请教您帮忙

**Comment by tingfengyinyue** [comment]
窗口提示：
Fatal Python error: initfsencoding: 
unable to load the file system codec zipimport.ZipImportError: 
can't find module' encodings

**Comment by saadi-tech** [comment]
Thank you sooooooooooooooooooooooooooooooooo much! :)

**Comment by jacobrafael08** [comment]
hi Masoud Rahimi 
please help me i cant seem to find spicy-extra-dll 
![image](https://user-images.githubusercontent.com/72881008/121677012-e8fbbc00-cae7-11eb-981a-6a55ad763d8e.png)

**Comment by jacobrafael08** [comment]
THANK YOU SO MUCH BRO!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

**Comment by divyangana009** [comment]
Error loading Python DLL 'C:\Users\91880\Desktop\application\python310.dll'. getting this error please help.

**Comment by masoudr** [comment]
@divyangana009 This is a basic error that I think might be related to PyInstaller itself. Have you tried freezing a simple script, I mean without `face_recognition`, and see if the executable can be run? Also, try to create a virtual environment and install the latest version of PyInstaller with `pip install -U PyInstaller`.

**Comment by SAUDIANO** [comment]
I've tried out what you suggested and my program runs with no errors in VSC but every time I convert it to .exe it simply runs so quickly that the terminal pops up so quick and close. I tried executing the exe file in Anaconda prompt and still giving me this error:

SyntaxError: Non-UTF-8 code starting with '\x90' in file C:\Users\PC\Desktop\VSworkSpace\FaceRecProject\dist\main\main.exe on line 1, but no encoding declared; see https://python.org/dev/peps/pep-0263/ for details

Even though I've added # -*- coding: utf-8 -*- at the first line of my code

![Screenshot 2022-07-23 165624](https://user-images.githubusercontent.com/109861513/180608198-860e280c-9663-416f-a8e7-e37a78164c48.jpg)

Please help me solving this issue

**Comment by masoudr** [comment]
@SAUDIANO I'm not sure what is the cause of the problem, but first, make sure that you run your executable directly from a standard CMD and not python [this](https://stackoverflow.com/questions/57391014/syntaxerror-non-utf-8-code-starting-with-x90-in-file-score-python-exe-on-l). Second, create a clean virtual environment with Conda and generate the executable in that env. Moreover, try to update the Pyinstaller to the latest version.

**Comment by sachinanalyst** [comment]
> Hi, You might want to freeze your script into a standalone executable to run on any system without the need of installing `python` or `face_recognition` and maybe you want to create a demo for your application and want to give it to someone else without giving your source code. Here is a simple tutorial to do that. I tested this method with `python3.6` on Windows 10, but I think you can get it working on Linux with a similar method.
> 
> 1. **Make sure** you have correctly installed both `face_recognition` and `dlib` and you see no error when importing them into your script.
> 2. **Make sure** your script works fine, and all its dependencies are right next to it, and you can run it fine with `python yourscript.py`.
> 3. Install [Pyinstaller ](https://github.com/pyinstaller/pyinstaller)with `pip`:
>    `pip install pyinstaller`
> 4. Create a new directory and move your python script and all dependencies into it. I call it `myproject` and `myscript.py`
> 5. Copy `face_recognition_models` and `scipy-extra-dll` from your python installed directory to your project directory.
> 6. Create an empty file called `<yourscriptname>.spec` like `myscript.spec` next to your python script.
> 7. Use below Pyinstaller spec file sample and edit some parts according to your needs: (I mark it with `<>` tag)
> 
> ```
> # -*- mode: python -*-
> 
> block_cipher = None
> 
> face_models = [
> ('.\\face_recognition_models\\models\\dlib_face_recognition_resnet_model_v1.dat', './face_recognition_models/models'),
> ('.\\face_recognition_models\\models\\mmod_human_face_detector.dat', './face_recognition_models/models'),
> ('.\\face_recognition_models\\models\\shape_predictor_5_face_landmarks.dat', './face_recognition_models/models'),
> ('.\\face_recognition_models\\models\\shape_predictor_68_face_landmarks.dat', './face_recognition_models/models'),
> ]
> 
> a = Analysis(['<your python script name.py>'],
>              pathex=['<path to working directory>'],
>              binaries=face_models,
>              datas=[],
>              hiddenimports=['scipy._lib.messagestream', 'scipy', 'scipy.signal', 'scipy.signal.bsplines', 'scipy.special', 'scipy.special._ufuncs_cxx',
>                             'scipy.linalg.cython_blas',
>                             'scipy.linalg.cython_lapack',
>                             'scipy.integrate',
>                             'scipy.integrate.quadrature',
>                             'scipy.integrate.odepack',
>                             'scipy.integrate._odepack',
>                             'scipy.integrate.quadpack',
>                             'scipy.integrate._quadpack',
>                             'scipy.integrate._ode',
>                             'scipy.integrate.vode',
>                             'scipy.integrate._dop', 'scipy._lib', 'scipy._build_utils','scipy.__config__',
>                             'scipy.integrate.lsoda', 'scipy.cluster', 'scipy.constants','scipy.fftpack','scipy.interpolate','scipy.io','scipy.linalg','scipy.misc','scipy.ndimage','scipy.odr','scipy.optimize','scipy.setup','scipy.sparse','scipy.spatial','scipy.special','scipy.stats','scipy.version'],
> 
>              hookspath=[],
>              runtime_hooks=[],
>              excludes=[],
>              win_no_prefer_redirects=False,
>              win_private_assemblies=False,
>              cipher=block_cipher)
> 
> a.datas += Tree('./scipy-extra-dll', prefix=None)
> 
> pyz = PYZ(a.pure, a.zipped_data,
>              cipher=block_cipher)
> exe = EXE(pyz,
>           a.scripts,
>           a.binaries,
>           a.zipfiles,
>           a.datas,
>           name='<your python script name>',
>           debug=False,
>           strip=False,
>           upx=True,
>           runtime_tmpdir=None,
>           console=True )
> ```
> 
> 8. Generate your executable with `python -m pyinstaller myscript.spec`
> 9. If you see no error then your executable can be found in `dist` directory.
> 10. Enjoy!
> 
> Thanks to @ageitgey and @davisking for their awesome work.



i use ultralytics then how to use
